stablehlo/reference/Ops.cpp

```diff
@@ -869,50 +893,39 @@ Tensor evalGatherOp(const Tensor &operand, const Tensor &startIndices,
                     ShapedType resultType) {
   Tensor result(resultType);
   Axes batchDims;
-  for (auto i = 0; i < result.getRank(); ++i)
-    if (!llvm::is_contained(offsetDims, i)) batchDims.push_back(i);
+  for (auto d : result.getAxes())
+    if (!llvm::is_contained(offsetDims, d)) batchDims.push_back(d);
 
   for (auto resultIt = result.index_begin(); resultIt != result.index_end();
        ++resultIt) {
     Index batchIndex;
-    for (auto batchDim : batchDims) batchIndex.push_back((*resultIt)[batchDim]);
-
-    Index sliceStartIndices(batchIndex);
-    Index sliceLimitIndices(Sizes(sliceStartIndices) + 1);
-    if (indexVectorDim < startIndices.getRank()) {
-      sliceStartIndices.insert(sliceStartIndices.begin() + indexVectorDim, 0);
-      sliceLimitIndices.insert(sliceLimitIndices.begin() + indexVectorDim,
-                               startIndices.getShape()[indexVectorDim]);
-    }
-    auto startIndexSlice =
-        evalSliceOp(startIndices, sliceStartIndices, sliceLimitIndices,
-                    Sizes(startIndices.getRank(), 1));
-    Index startIndex;
-    for (auto it = startIndexSlice.index_begin();
-         it != startIndexSlice.index_end(); ++it)
-      startIndex.push_back(
-          startIndexSlice.get(*it).getIntegerValue().getSExtValue());
+    for (auto d : batchDims) batchIndex.push_back((*resultIt)[d]);
+
+    Index startIndicesIndex = batchIndex;
+    if (indexVectorDim < startIndices.getRank())
+      startIndicesIndex.insert(startIndicesIndex.begin() + indexVectorDim,
+                               kColon);
+    Index startIndex = evalIndex(evalSliceOp(startIndices, startIndicesIndex));
 
     Index fullStartIndex(operand.getRank(), 0);
-    for (auto di = 0; di < operand.getRank(); ++di) {
-      if (llvm::find(startIndexMap, di) != startIndexMap.end()) {
-        auto dj = llvm::find(startIndexMap, di) - startIndexMap.begin();
-        fullStartIndex[di] = std::clamp(
-            startIndex[dj], 0L, operand.getShape()[di] - sliceSizes[di]);
```

Comment:
Why do we need a clamp here if we can just do `if (operandIndex < operand.getShape())` below? The alternative is obviously correct, whereas it wasn't obvious to me what to make of this code.

Reply:
You're right that we can just do `if (operandIndex < operand.getShape())` and it will still be a correct reference implementation as the spec says that the behavior is implementation defined if `operandIndex` is OOB. However, JAX assumes the clamping behavior, which I chose to implement (and to enable these tests).